### ISSUE TITLE
Deprecate `helpers.mkRaw` in favour of `lib.generators.mkLuaInline`

### ISSUE THREAD

**Issue by MattSturgeon** [nix-community/nixvim#1935]
- [ ] `helpers.lua.mkLua` should support nixpkg's "lua-inline" type.
- [ ] Documentation should recommend `mkLuaInline`
- [ ] Transition internal `__raw` to use `mkLuaInline`
- [ ] Add a deprecation warning to `helpers.mkRaw`

Note: there's now also a `lib.generators.toLua` in nixpkgs, however we currently can't use it, because it has no support for mixed list+attrs tables.

**Comment by khaneliman** [comment]
I'm a fan of transitioning away from a custom function to one maintained upstream. That would increase consistency with other repos and lower the tech debt with managing our own library functions.

**Comment by MattSturgeon** [comment]
I agree. I feel a little hesitant because we'll still be maintaining our own "unkeyed attrs" system, however I don't think this needs to be all or nothing straight away.

We can come up with an upstream solution to mixed lua tables separately. Thankfully this is less significant than raw/inline lua, because it is used far less often.

**Comment by MattSturgeon** [comment]
Something else to consider is how we show raw/inline values in the docs.

`lib.generators.toPretty` has support for a custom "pretty print" type (`allowPretty` iirc).

If that's enabled where the docs use `toPretty`, we could replace usage of `mkLuaInline` in defaultText&examples with a "forDocs" function that returns a pretty printer.

I think ideally, our examples should show the function _application_ not the resulting attrs.

Otherwise we'd just have to use `literalExpression`, e.g. `literalExpression "lib.generators.mkLuaInline { foo = false; }"`, but this is only used for _whole_ examples, not small parts of them.

**Comment by MattSturgeon** [comment]
We could have a `mkLuaInlineLiteral`, but since pretty-printers aren't supplied indentation information, we'd have to either work it out in advance or have no indentation.

```nix
mkLuaInlineLiteralIndented = indent: text: {
  __pretty =
    v:
    indent
    + "mkLuaInline "
    + generators.toPretty { inherit indent; } v;
  val = text;
};
```

It'd be nice if toPretty provided a way to call a pretty-printing function with additional information, such as the current indent and toPretty's args. Maybe if the pretty printer has `__prettyWith` instead of `__pretty`?

**Comment by GaetanLepage** [comment]
Thanks @MattSturgeon.
I agree with your migration plan. Doing it step by step looks great.

**Comment by MattSturgeon** [comment]
> With the conversation around #1935, #2035 will be another place we have a migration away from recommending `__raw`
> https://github.com/nix-community/nixvim/pull/2035#pullrequestreview-2244384155

I'm starting to think we should postpone publicly deprecating `mkRaw` until after 24.11... We have enough things for users to migrate already when theyre upgrading from 24.05->24.11 :sweat_smile:

**Comment by khaneliman** [comment]
> > With the conversation around #1935, #2035 will be another place we have a migration away from recommending `__raw`
> > [#2035 (review)](https://github.com/nix-community/nixvim/pull/2035#pullrequestreview-2244384155)
> 
> I'm starting to think we should postpone publicly deprecating `mkRaw` until after 24.11... We have enough things for users to migrate already when theyre upgrading from 24.05->24.11 😅

That makes sense actually.. there are a lot of deprecation notices throughout the codebase atm.
